trio/_threads.py

```diff
@@ -422,6 +427,9 @@ def from_thread_run_sync(fn, *args, trio_token=None):
           "foreign" thread, spawned using some other framework, and still want
           to enter Trio.
     """
+    if not (inspect.isfunction(fn) and not inspect.iscoroutinefunction(fn)):
```

Comment:
As discussed above, these won't tell you what you want. Better to wait and raise an error if the result of calling the function satisfies `inspect.iscoroutine`. (If it does, you'll also want to `close()` it before returning, so you don't get an extra "coroutine was never awaited" warning.)

Reply:
Just saw your comment on #1504, should I look into `Runner.spawn_impl()` for the "async fn or error" logic?

Reply:
That would be a useful extension if you feel like it! The relevant code is toward the top of spawn_impl(), with the comments about "Give good error for [...]"